docs/source/hooks.rst

```diff
@@ -0,0 +1,175 @@
+The :py:class:`~flow.FlowProject.operation_hooks` decorator tells **signac-flow** to run a hook (or set of hooks) when an operation reaches the specified trigger.
+
+The :py:class:`~flow.FlowProject.operation_hooks` decorator accepts objects as a function of the job operation (:py:class:`~flow.project.JobOperation`).
+
+An operation hook can be used to store basic information about the execution of a job operation in the job document.
+In the following example, when the test operation ``error_on_a_0`` raises an exception, the hook function ``store_error_to_doc`` executes.
+Otherwise, ``store_success_to_doc`` executes.
+
+.. code-block:: python
+
+    # project.py
+    from flow import FlowProject
+
+    class Project(FlowProject):
+        pass
+
+    def store_success_to_doc(operation_name, job):
+        job.doc.update({f'{operation_name}_success': True})
+
+    def store_error_to_doc(operation_name, error, job):
+        job.doc.update({f'{operation_name}_success': False})
+
+    @Project.operation
+    @Project.operation_hooks.on_success(store_success_to_doc)
+    @Project.operation_hooks.on_exception(store_error_to_doc)
+    @Project.post.isfile("result.txt")
+    def error_on_a_0(job):
+        if job.sp.a == 0:
+            # Have jobs with statepoint 'a' == 0 fail
+            raise RuntimeError("Cannot process jobs with a == 0.")
+
+    if __name__ == '__main__':
+        Project().main()
+
+
+If ``error_on_a_0`` is executed using ``python project.py run --operation error_on_a_0 --filter a 1``, the ``on_success`` hook trigger will run, and ``job.doc.error_on_a_0_success`` will be ``True``.
+
+If ``error_on_a_0`` is executed using ``python project.py run --operation error_on_a_0 --filter a 0``, a ``RuntimeError`` is raised.
+The ``on_exception`` hook trigger will run, and ``job.doc.error_on_a_0_success`` will be ``False``.
+
+
+.. _project-level-hooks:
+
+Project-Level Hooks
+===================
+
+It may be desirable to install the same hook or set of hooks for all operations in a project.
+In the following example FlowProject, the hook ``track_start_time`` is triggered when each operation starts.
+The hook appends the current time to a list in the job document that is named based on the name of the operation.
+
+ .. code-block:: python
+
+    from flow import FlowProject
+
+    class Project(FlowProject):
+        pass
+
+    @Project.operation
+    @Project.post.true('test_ran')
+    def do_operation(job):
+        job.doc.test_ran = True
+
+    @Project.operation
+    @Project.pre.after(do_operation)
+    @Project.post.false('test_ran')
+    def undo_operation(job):
+        job.doc.test_ran = False
+
+    def track_start_time(operation_name, job):
+        import time
+        current_time = time.strftime('%b %d, %Y at %l:%M:%S %p %Z')
+        doc_key = f'{operation_name}_start_times'
+        times = job.doc.get(doc_key, [])
+        times.append(current_time)
+        job.doc[doc_key] = times
```

Comment:
This is equivalent. Feel free to reject if you think the existing form is more readable.
```suggestion
        job.doc.setdefault(doc_key, [])
        job.doc[doc_key].append(current_time)
```